### ISSUE TITLE
Test terraform for AWS Autoscale Groups

### ISSUE THREAD

**Issue by anosulchik** [newcontext-oss/kitchen-terraform#142]
Hi there,

We'd like to test our terraform configuration with kitchen-terraform e.g. to make sure that EC2 instances are bootstrapped properly. For example, check that our specific service is up and running at newly created EC2 instance.

We have autoscaling group (ASG) defined resource defined in terraform configuration so ASG creates EC2 instances after `terraform apply`. At the same time, in order to make kitchen-terraform working we need to specify output that contains IP address or hostname for EC2 resources which needs to be tested by kitchen-terraform. We cannot do this in our setup.

Are there any options to configure kitchen-terraform to take IPs of EC2 instances which belong to specified autoscale group? Or at least disable requirement to specify output with IP to let kitchen-terraform working -- e.g. for cases when terraform configuration doesn't have EC2 instances but it should still be tested.

Thank you!!!

**Comment by ncs-alane** [comment]
@anosulchik: hello again! :laughing: 

The `:hostnames` subattribute of the [:groups configuration attribute](http://www.rubydoc.info/gems/kitchen-terraform/2.0.0/Kitchen/Terraform/ConfigAttribute/Groups) on the [kitchen-terraform verifier](http://www.rubydoc.info/gems/kitchen-terraform/2.0.0/Kitchen/Verifier/Terraform) can be omitted, which will cause the verifier to execute a group's InSpec controls on localhost. You could then use a Ruby gem like [awspec](https://rubygems.org/gems/awspec) write controls against the AWS API. 

Alternatively, if the IP addresses could be resolved in some manner after executing `kitchen converge` and before executing `kitchen verify`, it may still be possible to achieve the remote control execution, but more details about your particular project would be necessary.

**Comment by anosulchik** [comment]
@ncs-alane hello! :-)

I'll try to use latest kitchen-terraform and omit `:hostnames` subattribute, at this stage it would be enough to use kitchen-terraform to create aws resources and execute awspec tests (and destroy aws resources after tests were finished).

> it may still be possible to achieve the remote control execution

Can you give more information about it please? Do you mean terraform provisioner that can be run locally and collect IPs of EC2 instances? Probably it could be an option for us but not too elegant though :)

> but more details about your particular project would be necessary

Basically we have terraform template that we use to create ASG, cloudwatch alarms, security groups and so on. We're using own AMI for EC2 instances (this AMI is specified in ASG launch configuration). We would love to use kitchen-terraform to verify that there're specific services are running on EC2 instances created by ASG. To accomplish this, we need to configure kitchen-terraform with proper IPs which it can use to run tests. 

Thank you!

**Comment by ncs-alane** [comment]
There are a few ways that external data can be input when using kitchen-terraform, like using the [:variables configuration attribute](http://www.rubydoc.info/gems/kitchen-terraform/2.0.0/Kitchen/Terraform/ConfigAttribute/Variables) of the [kitchen-terraform driver](http://www.rubydoc.info/gems/kitchen-terraform/2.0.0/Kitchen/Driver/Terraform).

If there is some way for you to query AWS for a stable list of ASG IP addresses after the ASG has been provisioned, you could provide the IP addresses as a dynamic variable input to the Terraform module under test and then forward them to an output which could be used with the `:hostnames` configuration attribute.

This is an example of what that could look like, leveraging Test Kitchen's support for embedded Ruby in `.kitchen.yml`:

```yaml
driver:
  name: terraform
  variables:
    asg_ip_addresses: <%= TestModule.fetch_ip_addresses %>
```

**Comment by ncs-alane** [comment]
@anosulchik: did you have any success?

**Comment by anosulchik** [comment]
Not yet for this particular case, at the same time it's clear how to use suggested approach. I'm using it for getting list of EC2 instances and skip terminated ones. Thanks.
